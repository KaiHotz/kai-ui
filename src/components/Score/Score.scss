.ui-score {
  --ui-score--tick-color: var(--blue-3);

  position: relative;
  height: 80px;
  width: 80px;
  margin: auto;

  &__ticks {
    color: var(--ui-score--tick-color);
  }

  &__tick {
    transform-origin: 50px 50px;
  }

  &__value {
    position: absolute;
    top: 50%;
    left: 50%;
    transform: translate(-50%, -50%);
    color: var(--emphasis-primary);
    width: 50px;
    height: 50px;
    line-height: 50px;
    text-align: center;
    border-radius: 50%;
    border: 1px solid var(--blue-2);
  }
}
