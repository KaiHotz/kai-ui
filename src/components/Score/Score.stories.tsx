import React from 'react';
import type { Meta, StoryObj } from '@storybook/react';

import { Score } from './Score';

const meta: Meta<typeof Score> = {
  title: 'Components/Score',
  component: Score,
  argTypes: {
    testId: {
      control: false,
    },
  },
};

type Story = StoryObj<typeof Score>;
export const Default: Story = {
  render: (args) => {
    return <Score {...args} />;
  },
  args: {
    value: 30,
    ticks: 30,
    color: '',
  },
};

export default meta;
