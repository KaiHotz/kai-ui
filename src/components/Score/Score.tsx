import { FC, useMemo } from 'react';
import './Score.scss';

interface IScoreProps {
  value: string | number;
  ticks?: number;
  color?: string;
  testId?: string;
}
export const Score: FC<IScoreProps> = ({ ticks = 30, value, color, testId = 'ui-score' }) => {
  const ticksArr = useMemo(() => {
    const tickSize = 360 / ticks;

    return Array(ticks)
      .fill(null)
      .map((_, index) => tickSize * index);
  }, [ticks]);

  return (
    <div className="ui-score" style={{ ['--ui-score--tick-color' as 'color']: color }} data-testid={testId}>
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 100 100" className="ui-score__ticks">
        {ticksArr.map((degree, i) => (
          <rect
            key={`${i + 1}`}
            className="ui-score__tick"
            x="50px"
            y="50px"
            fill="currentColor"
            strokeLinecap="round"
            width="2px"
            height="9px"
            style={{ transform: `rotate(${degree}deg) translateY(40px)` }}
          />
        ))}
      </svg>
      <span className="ui-score__value" data-testid={`${testId}-value`}>
        {value}
      </span>
    </div>
  );
};
