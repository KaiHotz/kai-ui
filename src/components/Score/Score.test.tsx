import { describe, expect, it } from 'vitest';
import { render, screen } from '@testing-library/react';

import { Score } from './Score';

const baseProps = {
  value: 30,
};

describe('<Score />', () => {
  it('should render', () => {
    render(<Score {...baseProps} />);

    expect(screen.getByTestId('ui-score')).toMatchSnapshot();
  });
});
