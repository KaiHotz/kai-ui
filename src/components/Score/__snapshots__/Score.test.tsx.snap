// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<Score /> > should render 1`] = `
<div
  class="ui-score"
  data-testid="ui-score"
>
  <svg
    class="ui-score__ticks"
    viewBox="0 0 100 100"
    xmlns="http://www.w3.org/2000/svg"
  >
    <rect
      class="ui-score__tick"
      fill="currentColor"
      height="9px"
      stroke-linecap="round"
      style="transform: rotate(0deg) translateY(40px);"
      width="2px"
      x="50px"
      y="50px"
    />
    <rect
      class="ui-score__tick"
      fill="currentColor"
      height="9px"
      stroke-linecap="round"
      style="transform: rotate(12deg) translateY(40px);"
      width="2px"
      x="50px"
      y="50px"
    />
    <rect
      class="ui-score__tick"
      fill="currentColor"
      height="9px"
      stroke-linecap="round"
      style="transform: rotate(24deg) translateY(40px);"
      width="2px"
      x="50px"
      y="50px"
    />
    <rect
      class="ui-score__tick"
      fill="currentColor"
      height="9px"
      stroke-linecap="round"
      style="transform: rotate(36deg) translateY(40px);"
      width="2px"
      x="50px"
      y="50px"
    />
    <rect
      class="ui-score__tick"
      fill="currentColor"
      height="9px"
      stroke-linecap="round"
      style="transform: rotate(48deg) translateY(40px);"
      width="2px"
      x="50px"
      y="50px"
    />
    <rect
      class="ui-score__tick"
      fill="currentColor"
      height="9px"
      stroke-linecap="round"
      style="transform: rotate(60deg) translateY(40px);"
      width="2px"
      x="50px"
      y="50px"
    />
    <rect
      class="ui-score__tick"
      fill="currentColor"
      height="9px"
      stroke-linecap="round"
      style="transform: rotate(72deg) translateY(40px);"
      width="2px"
      x="50px"
      y="50px"
    />
    <rect
      class="ui-score__tick"
      fill="currentColor"
      height="9px"
      stroke-linecap="round"
      style="transform: rotate(84deg) translateY(40px);"
      width="2px"
      x="50px"
      y="50px"
    />
    <rect
      class="ui-score__tick"
      fill="currentColor"
      height="9px"
      stroke-linecap="round"
      style="transform: rotate(96deg) translateY(40px);"
      width="2px"
      x="50px"
      y="50px"
    />
    <rect
      class="ui-score__tick"
      fill="currentColor"
      height="9px"
      stroke-linecap="round"
      style="transform: rotate(108deg) translateY(40px);"
      width="2px"
      x="50px"
      y="50px"
    />
    <rect
      class="ui-score__tick"
      fill="currentColor"
      height="9px"
      stroke-linecap="round"
      style="transform: rotate(120deg) translateY(40px);"
      width="2px"
      x="50px"
      y="50px"
    />
    <rect
      class="ui-score__tick"
      fill="currentColor"
      height="9px"
      stroke-linecap="round"
      style="transform: rotate(132deg) translateY(40px);"
      width="2px"
      x="50px"
      y="50px"
    />
    <rect
      class="ui-score__tick"
      fill="currentColor"
      height="9px"
      stroke-linecap="round"
      style="transform: rotate(144deg) translateY(40px);"
      width="2px"
      x="50px"
      y="50px"
    />
    <rect
      class="ui-score__tick"
      fill="currentColor"
      height="9px"
      stroke-linecap="round"
      style="transform: rotate(156deg) translateY(40px);"
      width="2px"
      x="50px"
      y="50px"
    />
    <rect
      class="ui-score__tick"
      fill="currentColor"
      height="9px"
      stroke-linecap="round"
      style="transform: rotate(168deg) translateY(40px);"
      width="2px"
      x="50px"
      y="50px"
    />
    <rect
      class="ui-score__tick"
      fill="currentColor"
      height="9px"
      stroke-linecap="round"
      style="transform: rotate(180deg) translateY(40px);"
      width="2px"
      x="50px"
      y="50px"
    />
    <rect
      class="ui-score__tick"
      fill="currentColor"
      height="9px"
      stroke-linecap="round"
      style="transform: rotate(192deg) translateY(40px);"
      width="2px"
      x="50px"
      y="50px"
    />
    <rect
      class="ui-score__tick"
      fill="currentColor"
      height="9px"
      stroke-linecap="round"
      style="transform: rotate(204deg) translateY(40px);"
      width="2px"
      x="50px"
      y="50px"
    />
    <rect
      class="ui-score__tick"
      fill="currentColor"
      height="9px"
      stroke-linecap="round"
      style="transform: rotate(216deg) translateY(40px);"
      width="2px"
      x="50px"
      y="50px"
    />
    <rect
      class="ui-score__tick"
      fill="currentColor"
      height="9px"
      stroke-linecap="round"
      style="transform: rotate(228deg) translateY(40px);"
      width="2px"
      x="50px"
      y="50px"
    />
    <rect
      class="ui-score__tick"
      fill="currentColor"
      height="9px"
      stroke-linecap="round"
      style="transform: rotate(240deg) translateY(40px);"
      width="2px"
      x="50px"
      y="50px"
    />
    <rect
      class="ui-score__tick"
      fill="currentColor"
      height="9px"
      stroke-linecap="round"
      style="transform: rotate(252deg) translateY(40px);"
      width="2px"
      x="50px"
      y="50px"
    />
    <rect
      class="ui-score__tick"
      fill="currentColor"
      height="9px"
      stroke-linecap="round"
      style="transform: rotate(264deg) translateY(40px);"
      width="2px"
      x="50px"
      y="50px"
    />
    <rect
      class="ui-score__tick"
      fill="currentColor"
      height="9px"
      stroke-linecap="round"
      style="transform: rotate(276deg) translateY(40px);"
      width="2px"
      x="50px"
      y="50px"
    />
    <rect
      class="ui-score__tick"
      fill="currentColor"
      height="9px"
      stroke-linecap="round"
      style="transform: rotate(288deg) translateY(40px);"
      width="2px"
      x="50px"
      y="50px"
    />
    <rect
      class="ui-score__tick"
      fill="currentColor"
      height="9px"
      stroke-linecap="round"
      style="transform: rotate(300deg) translateY(40px);"
      width="2px"
      x="50px"
      y="50px"
    />
    <rect
      class="ui-score__tick"
      fill="currentColor"
      height="9px"
      stroke-linecap="round"
      style="transform: rotate(312deg) translateY(40px);"
      width="2px"
      x="50px"
      y="50px"
    />
    <rect
      class="ui-score__tick"
      fill="currentColor"
      height="9px"
      stroke-linecap="round"
      style="transform: rotate(324deg) translateY(40px);"
      width="2px"
      x="50px"
      y="50px"
    />
    <rect
      class="ui-score__tick"
      fill="currentColor"
      height="9px"
      stroke-linecap="round"
      style="transform: rotate(336deg) translateY(40px);"
      width="2px"
      x="50px"
      y="50px"
    />
    <rect
      class="ui-score__tick"
      fill="currentColor"
      height="9px"
      stroke-linecap="round"
      style="transform: rotate(348deg) translateY(40px);"
      width="2px"
      x="50px"
      y="50px"
    />
  </svg>
  <span
    class="ui-score__value"
    data-testid="ui-score-value"
  >
    30
  </span>
</div>
`;
